### PR TITLE
plugins/avante: init

### LINKED ISSUE #2096: [PLUGIN REQUEST] Avante
| Field    | Description |
|----------|-------------|
| Plugin   | `Avante`  | 
| Homepage | `https://github.com/yetone/avante.nvim`     | 
| Nixpkgs  | `false` |

### PR DESCRIPTION
**Pull request by GaetanLepage**
Add support for [avante.nvim](https://github.com/yetone/avante.nvim), an AI companion plugin.

cc @synthe102 for testing

Fixes #2096

### CODE REVIEW

File: plugins/by-name/avante/default.nix [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  inherit (lib.nixvim) defaultNullOpts mkNullOrOption' toLuaObject;
+in
+lib.nixvim.neovim-plugin.mkNeovimPlugin {
+  name = "avante";
+  originalName = "avante.nvim";
+  package = "avante-nvim";
+
+  maintainers = [ lib.maintainers.GaetanLepage ];
+
+  # TODO: remove when XXX reaches `nixos-unstable`
```
Comment (khaneliman) [inline plugins/by-name/avante/default.nix:12]: do you have an associate override update PR for this?

File: plugins/by-name/avante/default.nix [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      autojump = true;
+      list_opener = "copen";
+    };
+  };
+
+  callSetup = false;
+  extraConfig = cfg: {
+    extraConfigLua = ''
+      require('avante_lib').load()
+      require('avante').setup (${toLuaObject cfg.settings})
```
Comment (khaneliman) [inline plugins/by-name/avante/default.nix:92]: This is the kind of stuff that makes me want https://github.com/nix-community/nixvim/pull/1876 available. @traxys are you still interested in getting that in?

Comment (traxys) [reply plugins/by-name/avante/default.nix:92]: Yeah, need to finish addressing the feedback on it

File: plugins/by-name/avante/default.nix (hunk 1 above)
Comment (GaetanLepage) [reply plugins/by-name/avante/default.nix:12]: I am still looking for it...

Comment (GaetanLepage) [reply plugins/by-name/avante/default.nix:12]: Actually I was wrong. Everything is find and we shouldn't have to add the dependencies there.

File: plugins/by-name/avante/default.nix (hunk 2 above)
Comment (traxys) [reply plugins/by-name/avante/default.nix:85]: #1876 has been merged

Comment (GaetanLepage) [reply plugins/by-name/avante/default.nix:85]: In this case, I just add `require('avante_lib').load()` to `luaConfig.pre` ?

Comment (GaetanLepage) [reply plugins/by-name/avante/default.nix]: done

File: tests/test-sources/plugins/by-name/avante/default.nix [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        system_prompt = ''
+          You are an excellent programming expert.
+        '';
+        openai = {
+          endpoint = "https://api.openai.com/v1";
+          model = "gpt-4o";
+          timeout = 30000;
+          temperature = 0;
+          max_tokens = 4096;
+          "__rawKey__'local'" = false;
```
Comment (MattSturgeon) [inline tests/test-sources/plugins/by-name/avante/default.nix:24]: Is this rawkey needed? Shouldn't our `isLuaKeyword` logic already handle this case when rendering the lua table?

Comment (GaetanLepage) [reply tests/test-sources/plugins/by-name/avante/default.nix:24]: How should I use it ? It's quite new to me ? Simply `local = false;` ?

Comment (MattSturgeon) [reply tests/test-sources/plugins/by-name/avante/default.nix:24]: Yeah. That should work. If it doesn't then `toLua` needs some bug fixing 😁

**Review by MattSturgeon — APPROVED** [review]

**Comment by GaetanLepage** [comment]
@mergifyio queue

**Comment by GaetanLepage** [comment]
Thanks for the review !

**Comment by elythh** [comment]
Has anyone tested the diff bindings ? Can't get them to work

**Pull request merged**